### PR TITLE
Replace slow `reconstruct` from Parameters.jl with faster `@setfield` from Setfield.jl

### PR DESCRIPTION
**Pull request by torfjelde**
We currently use `Parameters.reconstruct` to "reconstruct"/update immutable objects with new values, e.g. `reconstruct(hamiltonian, adaptor)` to update the metric used. As described in the docstring of `Parameters.reconstruct`, this is not a performant impl (uses `Dict` under the hood to deal with the mappings), and it recommends using Setfield.jl instead. 

This PR attempts to move from `reconstruct` to `@setfield` where appropriate.

### CODE REVIEW

**Review by xukai92 — APPROVED** [review]
Looks good to me. As long as the CI passes this is ready to go.

File: Project.toml [hunk 1]
```diff
@@ -22,6 +23,7 @@ InplaceOps = "0.3"
 Parameters = "0.10, 0.11, 0.12"
```
[the remark this answers is not in the document]
Comment (torfjelde) [reply Project.toml:23]: Not if we replace it with UnPack, no:)

File: src/AdvancedHMC.jl [hunk 2]
```diff
@@ -7,7 +7,11 @@ using LinearAlgebra: Symmetric, UpperTriangular, mul!, ldiv!, dot, I, diag, chol
 using StatsFuns: logaddexp, logsumexp
 using Random: GLOBAL_RNG, AbstractRNG
 using ProgressMeter: ProgressMeter
-using Parameters: @unpack, reconstruct
```
[the remark this answers is not in the document]
Comment (torfjelde) [reply src/AdvancedHMC.jl:10]: Yep, we could do that:)

**Comment by torfjelde** [comment]
Could you check this now @xukai92 ?:)

**Review by xukai92 — APPROVED** [review]

**Comment by xukai92** [comment]
Looks great!

**Comment by xukai92** [comment]
This PR actually not longer make sure these internals immutable.
I'm now a bit concern whether or not that might introduce any side effects.
On the other side, can you confirm how much speed gain do we get, apart from the minimal Gaussian demo?

**Comment by torfjelde** [comment]
Am confused. This shouldn't introduce any mutations at all? That's the entire point of Setfield.jl, no?

**Comment by xukai92** [comment]
Oh, my bad. I was checking it and made a mistake in my code that made it look like `@set` mutates some field.

False alarm then!

**Comment by torfjelde** [comment]
Haha, good, good:)

**Pull request merged**